docs/proposed/psa-driver-interface.md

```diff
@@ -536,6 +537,37 @@ The core will only call `acme_export_public_key` on a private key. Drivers imple
 
 The core guarantees that the size of the output buffer (`data_size`) is sufficient to export any key with the given attributes. The driver must set `*data_length` to the exact size of the exported key.
 
+### Builtin keys for opaque drivers
+
+An opaque driver can be pre-provisioned from the factory with keys. These keys are never seen by the PSA Crypto core, but in order to be able to use these keys, they need to be declared to the core.
+
+An opaque driver can declare such keys in its JSON descriptor file. The declaration is an array of strings, where each string is the name of a preprovisioned key, e.g.:
+```
+{
+    "builtin_keys": ["pre_key_1", "pre_key_2"]
+}
+```
+
+The key names need to be compatible with C language variable naming constraints, i.e. they cannot contain spaces and cannot start with a digit.
+
+For each declared key, the respective driver needs to define and externally declare the corresponding PSA Crypto key attributes (`psa_key_attributes_t`). The name of this variable, for each key, is `{prefix}_builtin_{keyname}_attributes`, where `{prefix}` is the implementing driver's prefix (lowercase), and `{keyname}` is the lowercase name of the key from the `builtin_keys` array.
+
+E.g. for a key called `"identity"`, a driver called `"acme"` would need to expose:
+```
+extern psa_core_key_attributes_t acme_builtin_identity_attributes;
```

Comment:
I'm not convinced that an attributes object per key is the way to go. There's no portable way to construct a compile-time constant: how the information is represented in a `psa_key_attributes_t` object is implementation-specific and at most a matter of ABI compatibility, not API compatibility. Should this be a function that takes a key id (?) as argument and a pointer to an attribute structure to fill?

Reply:
That could indeed be another solution, which from a definition PoV is probably a cleaner solution, though at the tradeoff of requiring more code.